doc/monads/readers.lean

```diff
@@ -0,0 +1,199 @@
+> So technically if `return` is the last statement in a function it could be replaced with `pure <|`,
+but one could argue that `return` is still a little easier for most folks to read, just so long as
+you understand that `return` is doing more than other languages, it is also wrapping pure values in
+the monadic container type.
+
+> **Side note 2**: If the function `readerFunc3` also took some explicit arguments then you would have
+to write `(readerFunc3 args).run env` and this is a bit ugly, so Lean provides an infix operator
+`|>` that eliminiates those parens so you can write `readerFunc3 args |>.run env` and then you can
+chain multiple monadic actions like this `m1 args1 |>.run args2 |>.run args3` and this is the
+recommended style.  You will see this patten used heavily in Lean code.
+
+The `let env ← read` expression in `readerFunc1` unwraps the environment from the `ReaderM` so we
+can use it. Each type of monad might provide one or more extra functions like this, functions that
+become available only when you are in the context of that monad.
+
+Here the `readerFunc2` function uses the `bind` operator `>>=` just to show you that there are bind
+operations happening here.  The `readerFunc3` function uses the  `do` notation you learned about in
+[Monads](monads.lean.md) which hides that bind operation and can make the code look cleaner.
+
+The `do` notation with `let x ← readerFunc2` is also calling the `bind` function under the covers,
+so that you can access the unwrapped value `x` needed for the `toString x` conversion.
+
+The important difference here to the earlier code is that `readerFunc3` and `readerFunc2` no longer
+have an **explicit** Environment input parameter that needs to be passed along all the way to
+`readerFunc1`.  Instead, the `ReaderM` monad is taking care of that for you, which gives you the
+illusion of something like global context where the context is now available to all functions that use
+the `ReaderM` monad.
+
+The above code also introduces an important idea. Whenever you learn about a monad "X", there's
+often (but not always) a `run` function to execute that monad, and sometimes some additional
+functions like `read` that interact with the monad context.
+
+You might be wondering, how does the context actually move through the `ReaderM` monad? How can you
+add an input argument to a function by modifying it's return type?  There is a special command in
+the Lean interpreter that will show you the reduced Types:
+-/
+#reduce ReaderM Environment String   -- Environment → String
+/-!
+And you can see here that this type is actually a function!  It's a function that takes an
+`Environment` as input and returns a `String`.
+
+Now, remember in Lean that a function that takes argument of type `Nat` and returns a `String` like
+`def f (a : Nat) : String` is the same as this function `def f : Nat → String`.  These are
+definitially equal.  Well this is being used by the `ReaderM` Monad to add an input argument to
```

Comment:
You could even say "exactly equal [as types]" here. I didn't check if you discuss defeq in other places.

Reply:
fixed.

Reply:
Fixed, I wrote this instead: 
> These are exactly equal as types, which means they are "definitionally equal"

Reply:
That makes it sound like that's the definition of definitional equality, which is misleading at best. I don't know why you want to reference that here, it doesn't seem relevant at all.

Reply:
I see, well I was just using the language Leo gave me earlier, but I dropped it since it seems to be confusing things: https://github.com/lovettchris/lean4/commit/4b9d73814a625e08135719672cbf611134a8e472

Reply:
@lovettchris I used definitionally equality in the `ReaderM Environment String` vs `Environment -> String` issue.

Reply:
Right, but I agree with Sebastian it's an advanced topic better left to TPIL book, so probably doesn't belong here.